include/wlr/types/wlr_relative_pointer_v1.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * This an unstable interface of wlroots. No guarantees are made regarding the
+ * future consistency of this API.
+ */
+#ifndef WLR_USE_UNSTABLE
+#error "Add -DWLR_USE_UNSTABLE to enable unstable wlroots features"
+#endif
+
+#ifndef WLR_TYPES_WLR_RELATIVE_POINTER_V1_H
+#define WLR_TYPES_WLR_RELATIVE_POINTER_V1_H
+
+#include <wayland-server.h>
+
+
+/**
+ * This protocol specifies a set of interfaces used for making clients able to
+ * receive relative pointer events not obstructed by barriers (such as the
+ * monitor edge or other pointer barriers).
+ */
+
+
+/**
+ * A global interface used for getting the relative pointer object for a given
+ * pointer.
+ *
+ * Signals:
+ * - destroy -> struct wlr_relative_pointer_manager_v1 *manager
+ *   :: the manager was destroyed
+ * - new_relative_pointer -> struct wlr_relative_pointer_v1 *relative_pointer
+ *   :: a new relative_pointer was created
+ */
+
+struct wlr_relative_pointer_manager_v1 {
+	struct wl_list resources;
+	struct wl_global *global;
+
+	struct {
+		struct wl_signal destroy;
+		struct wl_signal new_relative_pointer;
+	} events;
+
+	void *data;
+};
+
+
+/**
+ * A wp_relative_pointer object is an extension to the wl_pointer interface
+ * used for emitting relative pointer events. It shares the same focus as
+ * wl_pointer objects of the same seat and will only emit events when it has
+ * focus.
+ *
+ * Signals:
+ * - destroy -> struct wlr_relative_pointer_v1 *relative_pointer
+ *   :: the relative_pointer was destroyed
+ */
+
+struct wlr_relative_pointer_v1 {
+	struct wl_client *client;
+	struct wl_resource *resource;
+	struct wl_pointer *pointer;
```

Comment:
`wl_pointer` is a client-side struct, it shouldn't be used here?

Reply:
What should be the correct behavior [here](https://github.com/swaywm/wlroots/pull/1274/files#diff-4aa3486289ec9789afb8643384856f95R326)? To send the frame event to every pointer resource?

I thought the `wl_pointer` needed to be stored to compare against events/pointer resources, but if I'm not wrong, every seat has a unique pointer, right? So, `wlr_seat_client_from_pointer_resource` should be enough?

Reply:
`wl_pointer` is a client struct that wraps `wl_proxy`. On the server-side we're using `wl_resource`, we shouldn't cast to `wl_pointer`.

This line of code should be `struct wl_resource *pointer`. An even better solution would be to store a pointer to the seat client (`struct wlr_seat_client *seat_client`).

Reply:
I see. Could you check and tell me what the correct behavior should be when sending frame events in the file I linked above? If sending it to every pointer resource of the same seat-client suffices, there's no need to store anything.

Reply:
Ah, right, each seat client can have multiple pointer resources, so storing `wlr_seat_client` won't do, we need to store the `wl_resource` and send the frame event only to this one.

Reply:
Done. Exposed `wlr_relative_pointer_v1_from_resource` as well, for the frame event handling.